File: core/java/src/main/java/hara/lang/protocol/IMatch.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(
    namespace = "std.protocol.imatch",
    name = "IMatch",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IMatch {
  @HaraMethod(value = "match-value", arity = 2)
  Object matchValue(Object value);
}

File: core/rust/src/lang/protocol/imatch.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.imatch",
    name = "IMatch",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IMatch {
    type Value;
    type Output;

    #[hara_method(value = "match-value", arity = 2)]
    fn match_value(&self, value: Self::Value) -> Self::Output;
}
